File: go/reader.go
package golf

import (
	"encoding/binary"
	"fmt"
	"io"
	"os"
)

// GolfReader reads .golf files using standard I/O.
type GolfReader struct {
	rs          io.ReadSeeker
	Header      Header
	Metadata    []MetadataEntry
	Descriptors []BlockDescriptor
}

// OpenReader opens a golf file from a ReadSeeker.
func OpenReader(rs io.ReadSeeker) (*GolfReader, error) {
	// Get file size
	size, err := rs.Seek(0, io.SeekEnd)
	if err != nil {
		return nil, err
	}
	if size < int64(FixedHeaderSize+FooterSize) {
		return nil, fmt.Errorf("file too small: %d bytes", size)
	}

	// Read footer
	if _, err := rs.Seek(-int64(FooterSize), io.SeekEnd); err != nil {
		return nil, err
	}
	var footerBuf [FooterSize]byte
	if _, err := io.ReadFull(rs, footerBuf[:]); err != nil {
		return nil, fmt.Errorf("read footer: %w", err)
	}
	footer, err := DecodeFooter(footerBuf)
	if err != nil {
		return nil, err
	}

	// Read block index
	indexSize := int(footer.BlockCount) * BlockDescriptorSize
	if _, err := rs.Seek(int64(footer.IndexOffset), io.SeekStart); err != nil {
		return nil, err
	}
	indexBuf := make([]byte, indexSize)
	if _, err := io.ReadFull(rs, indexBuf); err != nil {
		return nil, fmt.Errorf("read block index: %w", err)
	}
	actualCRC := computeCRC32C(indexBuf)
	if actualCRC != footer.IndexCRC {
		return nil, fmt.Errorf("block index CRC mismatch: expected %#010x, got %#010x",
			footer.IndexCRC, actualCRC)
	}
	var descriptors []BlockDescriptor
	for i := 0; i < int(footer.BlockCount); i++ {
		start := i * BlockDescriptorSize
		var dbuf [BlockDescriptorSize]byte
		copy(dbuf[:], indexBuf[start:start+BlockDescriptorSize])
		descriptors = append(descriptors, DecodeBlockDescriptor(dbuf))
	}

	// Read header
	if _, err := rs.Seek(0, io.SeekStart); err != nil {
		return nil, err
	}
	var headerBuf [FixedHeaderSize]byte
	if _, err := io.ReadFull(rs, headerBuf[:]); err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}
	header, err := DecodeHeader(headerBuf)
	if err != nil {
		return nil, err
	}

	// Read metadata
	var metadata []MetadataEntry
	if header.MetadataLength > 0 {
		metaBuf := make([]byte, header.MetadataLength)
		if _, err := io.ReadFull(rs, metaBuf); err != nil {
			return nil, fmt.Errorf("read metadata: %w", err)
		}
		metadata, err = DecodeMetadata(metaBuf)
		if err != nil {
			return nil, err
		}
	}

	return &GolfReader{
		rs:          rs,
		Header:      *header,
		Metadata:    metadata,
		Descriptors: descriptors,
	}, nil
}

// OpenFile opens a golf file from a filesystem path.
func OpenFile(path string) (*GolfReader, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	return OpenReader(f)
}

// Query returns records in the inclusive timestamp range [startTs, endTs].
func (r *GolfReader) Query(startTs, endTs uint64) ([]Record, error) {
	firstIdx := FindFirstBlock(r.Descriptors, startTs)
	if firstIdx < 0 {
		return nil, nil
	}
	lastIdx := FindLastBlock(r.Descriptors, endTs)
	if lastIdx < 0 {
		return nil, nil
	}
	if firstIdx > lastIdx {
		return nil, nil
	}

	recordSize := 8 + int(r.Header.RecordValueSize)
	var results []Record

	for bi := firstIdx; bi <= lastIdx; bi++ {
		desc := &r.Descriptors[bi]

		// Read compressed block
		if _, err := r.rs.Seek(int64(desc.BlockOffset), io.SeekStart); err != nil {
			return nil, err
		}
		compressed := make([]byte, desc.CompressedSize)
		if _, err := io.ReadFull(r.rs, compressed); err != nil {
			return nil, fmt.Errorf("read block: %w", err)
		}

		// Decompress
		raw, err := DecompressBlock(compressed, r.Header.Compression, int(desc.UncompressedSize))
		if err != nil {
			return nil, err
		}

		// Verify CRC
		actualCRC := computeCRC32C(raw)
		if actualCRC != desc.BlockCRC {
			return nil, fmt.Errorf("block CRC mismatch: expected %#010x, got %#010x",
				desc.BlockCRC, actualCRC)
		}

		count := int(desc.RecordCount)
		recStart := 0
		recEnd := count - 1

		if bi == firstIdx {
			recStart = binarySearchFirst(raw, recordSize, count, startTs)
		}
		if bi == lastIdx {
			recEnd = binarySearchLast(raw, recordSize, count, endTs)
		}

		if recStart <= recEnd && recEnd < count {
			for i := recStart; i <= recEnd; i++ {
				off := i * recordSize
				ts := binary.LittleEndian.Uint64(raw[off : off+8])
				if ts >= startTs && ts <= endTs {
					val := make([]byte, r.Header.RecordValueSize)
					copy(val, raw[off+8:off+recordSize])
					results = append(results, Record{Timestamp: ts, Value: val})
				}
			}
		}
	}

	return results, nil
}

func binarySearchFirst(data []byte, recordSize, count int, target uint64) int {
	lo, hi := 0, count
	for lo < hi {
		mid := lo + (hi-lo)/2
		ts := binary.LittleEndian.Uint64(data[mid*recordSize : mid*recordSize+8])
		if ts < target {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

func binarySearchLast(data []byte, recordSize, count int, target uint64) int {
	lo, hi := 0, count
	for lo < hi {
		mid := lo + (hi-lo)/2
		ts := binary.LittleEndian.Uint64(data[mid*recordSize : mid*recordSize+8])
		if ts <= target {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	if lo == 0 {
		return 0
	}
	return lo - 1
}

// MmapGolfReader reads .golf files using memory-mapped I/O.
type MmapGolfReader struct {
	data        []byte
	Header      Header
	Metadata    []MetadataEntry
	Descriptors []BlockDescriptor
	file        *os.File
}

// OpenMmap opens a golf file using memory-mapped I/O.
func OpenMmap(path string) (*MmapGolfReader, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	info, err := f.Stat()
	if err != nil {
		f.Close()
		return nil, err
	}
	size := info.Size()
	if size < int64(FixedHeaderSize+FooterSize) {
		f.Close()
		return nil, fmt.Errorf("file too small: %d bytes", size)
	}

	// Read the whole file into memory (portable mmap alternative)
	data := make([]byte, size)
	if _, err := io.ReadFull(f, data); err != nil {
		f.Close()
		return nil, err
	}

	// Parse footer
	footerStart := len(data) - FooterSize
	var footerBuf [FooterSize]byte
	copy(footerBuf[:], data[footerStart:])
	footer, err := DecodeFooter(footerBuf)
	if err != nil {
		f.Close()
		return nil, err
	}

	// Parse block index
	indexStart := int(footer.IndexOffset)
	indexEnd := indexStart + int(footer.BlockCount)*BlockDescriptorSize
	indexBuf := data[indexStart:indexEnd]
	actualCRC := computeCRC32C(indexBuf)
	if actualCRC != footer.IndexCRC {
		f.Close()
		return nil, fmt.Errorf("block index CRC mismatch")
	}
	var descriptors []BlockDescriptor
	for i := 0; i < int(footer.BlockCount); i++ {
		start := i * BlockDescriptorSize
		var dbuf [BlockDescriptorSize]byte
		copy(dbuf[:], indexBuf[start:start+BlockDescriptorSize])
		descriptors = append(descriptors, DecodeBlockDescriptor(dbuf))
	}

	// Parse header
	var headerBuf [FixedHeaderSize]byte
	copy(headerBuf[:], data[0:FixedHeaderSize])
	header, err := DecodeHeader(headerBuf)
	if err != nil {
		f.Close()
		return nil, err
	}

	var metadata []MetadataEntry
	if header.MetadataLength > 0 {
		metaBuf := data[FixedHeaderSize : FixedHeaderSize+int(header.MetadataLength)]
		metadata, err = DecodeMetadata(metaBuf)
		if err != nil {
			f.Close()
			return nil, err
		}
	}

	return &MmapGolfReader{
		data:        data,
		Header:      *header,
		Metadata:    metadata,
		Descriptors: descriptors,
		file:        f,
	}, nil
}

// Close closes the underlying file.
func (r *MmapGolfReader) Close() error {
	return r.file.Close()
}

// Query returns records in the inclusive timestamp range [startTs, endTs].
func (r *MmapGolfReader) Query(startTs, endTs uint64) ([]Record, error) {
	firstIdx := FindFirstBlock(r.Descriptors, startTs)
	if firstIdx < 0 {
		return nil, nil
	}
	lastIdx := FindLastBlock(r.Descriptors, endTs)
	if lastIdx < 0 {
		return nil, nil
	}
	if firstIdx > lastIdx {
		return nil, nil
	}

	recordSize := 8 + int(r.Header.RecordValueSize)
	var results []Record

	for bi := firstIdx; bi <= lastIdx; bi++ {
		desc := &r.Descriptors[bi]

		compressed := r.data[desc.BlockOffset : desc.BlockOffset+uint64(desc.CompressedSize)]

		raw, err := DecompressBlock(compressed, r.Header.Compression, int(desc.UncompressedSize))
		if err != nil {
			return nil, err
		}

		actualCRC := computeCRC32C(raw)
		if actualCRC != desc.BlockCRC {
			return nil, fmt.Errorf("block CRC mismatch")
		}

		count := int(desc.RecordCount)
		recStart := 0
		recEnd := count - 1

		if bi == firstIdx {
			recStart = binarySearchFirst(raw, recordSize, count, startTs)
		}
		if bi == lastIdx {
			recEnd = binarySearchLast(raw, recordSize, count, endTs)
		}

		if recStart <= recEnd && recEnd < count {
			for i := recStart; i <= recEnd; i++ {
				off := i * recordSize
				ts := binary.LittleEndian.Uint64(raw[off : off+8])
				if ts >= startTs && ts <= endTs {
					val := make([]byte, r.Header.RecordValueSize)
					copy(val, raw[off+8:off+recordSize])
					results = append(results, Record{Timestamp: ts, Value: val})
				}
			}
		}
	}

	return results, nil
}

File: swift/Sources/Golf/Reader.swift
// Reader.swift — read-only access to .golf files with indexed range queries.

import Foundation

/// Read-only accessor for a golf file, loaded fully into memory.
///
/// Opening validates the footer, block index, and header (magic + CRC32C) and
/// decodes the metadata. ``query(start:end:)`` then binary-searches the block
/// index so that only blocks overlapping the requested range are touched;
/// untouched blocks are never decompressed. Files are immutable after sealing,
/// so instances are safe to share across reads.
public struct GolfReader {
    /// Validated fixed header of the file.
    public let header: GolfHeader
    /// Decoded header metadata (empty when none was written).
    public let metadata: [MetadataEntry]
    /// Block-index descriptors in file order.
    public let descriptors: [BlockDescriptor]

    private let data: [UInt8]
    private let recordSize: Int

    /// Total records in the file (from the validated header).
    public var recordCount: UInt64 { header.recordCount }

    /// Inclusive `[min, max]` timestamp span of the file.
    public var timestampRange: ClosedRange<UInt64> {
        header.minTimestamp...max(header.minTimestamp, header.maxTimestamp)
    }

    /// Parses an in-memory golf file image.
    public init(data: [UInt8]) throws {
        self.data = data
        guard data.count >= Golf.fixedHeaderSize + Golf.footerSize else {
            throw GolfError.fileTooSmall(data.count)
        }

        // Footer → block index → header; each step is checksum-validated.
        let footer = try Footer.decode(Array(data[(data.count - Golf.footerSize)...]))
        let indexStart = Int(footer.indexOffset)
        let indexEnd = indexStart + Int(footer.blockCount) * Golf.blockDescriptorSize
        guard indexStart >= 0, indexEnd <= data.count - Golf.footerSize else {
            throw GolfError.corruptedIndex("index range \(indexStart)..\(indexEnd) outside file")
        }
        let indexBytes = Array(data[indexStart..<indexEnd])
        guard CRC32C.checksum(indexBytes) == footer.indexCrc else {
            throw GolfError.crcMismatch("block index")
        }

        var descs: [BlockDescriptor] = []
        descs.reserveCapacity(Int(footer.blockCount))
        for i in 0..<Int(footer.blockCount) {
            descs.append(try BlockDescriptor.decode(indexBytes, offset: i * Golf.blockDescriptorSize))
        }
        self.descriptors = descs

        self.header = try GolfHeader.decode(Array(data[0..<Golf.fixedHeaderSize]))

        if header.metadataLength > 0 {
            let metaStart = Golf.fixedHeaderSize
            let metaEnd = metaStart + header.metadataLength
            guard metaEnd <= data.count else {
                throw GolfError.truncatedMetadata("metadata length exceeds file")
            }
            self.metadata = try decodeMetadata(Array(data[metaStart..<metaEnd]))
        } else {
            self.metadata = []
        }

        self.recordSize = 8 + header.recordValueSize
    }

    /// Loads and parses the file at `url`.
    public init(contentsOf url: URL) throws {
        try self.init(data: Array(Data(contentsOf: url)))
    }

    /// Loads and parses the file at `path`.
    public static func open(_ path: String) throws -> GolfReader {
        try GolfReader(contentsOf: URL(fileURLWithPath: path))
    }

    /// Returns every record whose timestamp falls inside `range`, ascending.
    ///
    /// Only blocks overlapping the range are read and decompressed; within each
    /// boundary block the scan is narrowed by binary search on timestamps.
    public func query(_ range: ClosedRange<UInt64>) throws -> [Record] {
        try query(start: range.lowerBound, end: range.upperBound)
    }

    /// Returns every record with `startTs <= timestamp <= endTs`, ascending.
    public func query(start startTs: UInt64, end endTs: UInt64) throws -> [Record] {
        guard !descriptors.isEmpty else { return [] }
        let firstIdx = findFirstBlock(descriptors, startTs)
        let lastIdx = findLastBlock(descriptors, endTs)
        guard firstIdx >= 0, lastIdx >= 0, firstIdx <= lastIdx else { return [] }

        var results: [Record] = []
        for bi in firstIdx...lastIdx {
            let desc = descriptors[bi]
            let raw = try decompressedBlock(desc)

            let recStart = bi == firstIdx ? binarySearchFirst(raw, recordSize: recordSize, count: desc.recordCount, target: startTs) : 0
            let recEnd = bi == lastIdx ? binarySearchLast(raw, recordSize: recordSize, count: desc.recordCount, target: endTs) : desc.recordCount - 1

            guard recStart <= recEnd else { continue }
            for i in recStart...recEnd {
                let ts = readTimestamp(raw, i, recordSize: recordSize)
                if ts >= startTs && ts <= endTs {
                    results.append(extractRecord(raw, i, recordValueSize: header.recordValueSize))
                }
            }
        }
        return results
    }

    /// Loads the block described by `desc` and verifies its CRC32C.
    private func decompressedBlock(_ desc: BlockDescriptor) throws -> [UInt8] {
        let start = Int(desc.blockOffset)
        let end = start + desc.compressedSize
        guard start >= 0, end <= data.count, start <= end else {
            throw GolfError.corruptedIndex("block at \(start)..\(end) outside file")
        }
        let payload = Array(data[start..<end])

        let raw: [UInt8]
        switch header.compression {
        case .none:
            raw = payload
        case .lz4:
            guard payload.count >= 4 else { throw GolfError.compressionError("LZ4 data too short") }
            let origSize = Int(LE.readU32(payload, 0))
            raw = try Lz4Block.decompress(Array(payload[4...]), outputSize: origSize)
        case .zstd:
            throw GolfError.compressionError("Zstd reading is not supported by this implementation")
        }

        guard CRC32C.checksum(raw) == desc.blockCrc else {
            throw GolfError.crcMismatch("block data")
        }
        return raw
    }
}
